File: Python/rat in a maze.py
def rat_in_maze(maze):
    """
    Rat in a Maze:
    - Find all paths from (0, 0) to (n-1, n-1) only moving Down, Left, Right, Up.
    - Only move through 1s (open path), avoid 0s (walls).
    - Uses backtracking to explore all directions.
    """
    n = len(maze)
    result = []
    directions = [('D', 1, 0), ('L', 0, -1), ('R', 0, 1), ('U', -1, 0)]
    visited = [[False]*n for _ in range(n)]

    def is_valid(x, y):
        return 0 <= x < n and 0 <= y < n and maze[x][y] == 1 and not visited[x][y]

    def backtrack(x, y, path):
        if x == n-1 and y == n-1:
            result.append(path)
            return
        for dir_symbol, dx, dy in directions:
            nx, ny = x + dx, y + dy
            if is_valid(nx, ny):
                visited[x][y] = True
                backtrack(nx, ny, path + dir_symbol)
                visited[x][y] = False  # Backtrack

    if maze[0][0] == 1:
        backtrack(0, 0, "")

    return result

# ✅ Example Maze
maze = [
    [1, 0, 0, 0],
    [1, 1, 0, 1],
    [0, 1, 0, 0],
    [1, 1, 1, 1]
]
print("Maze Paths:", rat_in_maze(maze))

File: C++/rat in a maze.cpp
#include <iostream>
#include <vector>
using namespace std;

void solveMaze(int x, int y, vector<vector<int>>& maze, vector<vector<int>>& visited,
               string path, vector<string>& paths, int n) {
    // Boundary + visited + wall check
    if (x < 0 || y < 0 || x >= n || y >= n || maze[x][y] == 0 || visited[x][y]) return;

    // Reached destination
    if (x == n - 1 && y == n - 1) {
        paths.push_back(path);
        return;
    }

    visited[x][y] = 1;

    // Move Down
    solveMaze(x + 1, y, maze, visited, path + 'D', paths, n);
    // Move Left
    solveMaze(x, y - 1, maze, visited, path + 'L', paths, n);
    // Move Right
    solveMaze(x, y + 1, maze, visited, path + 'R', paths, n);
    // Move Up
    solveMaze(x - 1, y, maze, visited, path + 'U', paths, n);

    visited[x][y] = 0;  // Backtrack
}

// ✅ Use-case
int main() {
    vector<vector<int>> maze = {
        {1, 0, 0, 0},
        {1, 1, 0, 1},
        {0, 1, 0, 0},
        {1, 1, 1, 1}
    };
    int n = 4;
    vector<vector<int>> visited(n, vector<int>(n, 0));
    vector<string> paths;

    if (maze[0][0] == 1)
        solveMaze(0, 0, maze, visited, "", paths, n);

    for (string& path : paths)
        cout << path << endl;  // All valid paths
}
